src/core/comm/coll.cc

```diff
@@ -0,0 +1,451 @@
+#else
+  printf("Not implemented\n");
+  assert(0);
+#endif
+}
+
+// called from main thread
+int collInit(int argc, char* argv[])
+{
+  current_unique_id = 0;
+#if defined(LEGATE_USE_GASNET)
+  int provided, res, init_flag = 0;
+  MPI_Initialized(&init_flag);
+  if (!init_flag) {
+    res = MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+    assert(res == MPI_SUCCESS);
+  } else {
+    log_coll.print(
+      "Warning: MPI has been initialized by others, make sure MPI is initialized with "
+      "MPI_THREAD_MULTIPLE");
+  }
+  mpi_comms.resize(MAX_NB_COMMS, MPI_COMM_NULL);
+  for (int i = 0; i < MAX_NB_COMMS; i++) {
+    res = MPI_Comm_dup(MPI_COMM_WORLD, &mpi_comms[i]);
+    assert(res == MPI_SUCCESS);
+  }
+#else
+  thread_comms.resize(MAX_NB_COMMS);
+  for (int i = 0; i < MAX_NB_COMMS; i++) {
+    thread_comms[i].ready_flag = false;
+    thread_comms[i].buffers    = nullptr;
+    thread_comms[i].displs     = nullptr;
+  }
+#endif
+  coll_inited = true;
+  return CollSuccess;
+}
+
+int collFinalize(void)
+{
+  assert(coll_inited == true);
+  coll_inited = false;
+#if defined(LEGATE_USE_GASNET)
+  int res;
+  for (int i = 0; i < MAX_NB_COMMS; i++) {
+    res = MPI_Comm_free(&mpi_comms[i]);
+    assert(res == MPI_SUCCESS);
+  }
+  mpi_comms.clear();
+  return MPI_Finalize();
+#else
+  for (int i = 0; i < MAX_NB_COMMS; i++) { assert(thread_comms[i].ready_flag == false); }
+  thread_comms.clear();
+  return CollSuccess;
+#endif
+}
+
+int collGetUniqueId(int* id)
+{
+  *id = current_unique_id;
+  current_unique_id++;
+  assert(current_unique_id <= MAX_NB_COMMS);
+  return CollSuccess;
+}
+
+#if defined(LEGATE_USE_GASNET)
+MPI_Datatype collDtypeToMPIDtype(CollDataType dtype)
+{
+  if (dtype == CollDataType::CollInt8) {
+    return MPI_INT8_T;
+  } else if (dtype == CollDataType::CollChar) {
+    return MPI_CHAR;
+  } else if (dtype == CollDataType::CollUint8) {
+    return MPI_UINT8_T;
+  } else if (dtype == CollDataType::CollInt) {
+    return MPI_INT;
+  } else if (dtype == CollDataType::CollUint32) {
+    return MPI_UINT32_T;
+  } else if (dtype == CollDataType::CollInt64) {
+    return MPI_INT64_T;
+  } else if (dtype == CollDataType::CollUint64) {
+    return MPI_UINT64_T;
+  } else if (dtype == CollDataType::CollFloat) {
+    return MPI_FLOAT;
+  } else if (dtype == CollDataType::CollDouble) {
+    return MPI_DOUBLE;
+  } else {
+    assert(0);
+    return MPI_BYTE;
+  }
+}
+
+int collGenerateAlltoallTag(int rank1, int rank2, CollComm global_comm)
+{
+  // tag: seg idx + rank_idx + tag
+  // int send_tag = ((sendto_global_rank * 10000 + global_rank) * 10 + ALLTOALL_TAG) * 10 +
+  // global_comm->unique_id; // which dst seg it sends to (in dst rank) int recv_tag = ((global_rank
+  // * 10000 + recvfrom_global_rank) * 10 + ALLTOALL_TAG) * 10 + global_comm->unique_id; // idx of
+  // current seg we are receving (in src/my rank)
+#if 1
+  int tag = (rank1 * 10000 + rank2) * CollTag::MAX_TAG + CollTag::ALLTOALL_TAG;
```

Comment:
two comments: 1) I suggest you use shift and bitwise operators for the encoding here and other places. 2) please give a meaningful name to the constant `10000`.

Reply:
partial fixed.